homeassistant/components/sensor/fritzbox_callmonitor.py

```diff
@@ -160,13 +206,76 @@ def _parse(self, line):
                    "device": line[6],
                    "initiated": isotime}
+            att["to_name"] = self._sensor.number_to_name(att["to"])
             self._sensor.set_attributes(att)
         elif line[1] == "CONNECT":
             self._sensor.set_state(VALUE_CONNECT)
             att = {"with": line[4], "device": [3], "accepted": isotime}
+            att["with_name"] = self._sensor.number_to_name(att["with"])
             self._sensor.set_attributes(att)
         elif line[1] == "DISCONNECT":
             self._sensor.set_state(VALUE_DISCONNECT)
             att = {"duration": line[3], "closed": isotime}
             self._sensor.set_attributes(att)
-        self._sensor.schedule_update_ha_state()
+        self._sensor.update_ha_state()
+
+
+class FritzBoxPhonebook(object):
+    """This connects to a FritzBox router and downloads its phone book."""
+
+    def __init__(self, host, port, username, password,
+                 phonebook_id=0, prefixes=None):
+        """Initialize the class."""
+        self.host = host
+        self.username = username
+        self.password = password
+        self.port = port
+        self.phonebook_id = phonebook_id
+        self.phonebook_dict = None
+        self.number_dict = None
+        if prefixes is not None:
+            self.prefixes = prefixes
+        else:
+            self.prefixes = []
+
+        # pylint: disable=import-error
+        import fritzconnection as fc
+        # Establish a connection to the FRITZ!Box.
+        self.fph = fc.FritzPhonebook(address=self.host,
+                                     user=self.username,
+                                     password=self.password)
+
+        if self.phonebook_id not in self.fph.list_phonebooks:
+            raise ValueError("Phonebook with this ID not found.")
+
+        self.update_phonebook()
+
+    @Throttle(MIN_TIME_PHONEBOOK_UPDATE)
+    def update_phonebook(self):
+        """Update the phone book dictionary."""
+        self.phonebook_dict = self.fph.get_all_names(self.phonebook_id)
+        self.number_dict = {re.sub(r'[^\d\+]', '', nr): name
+                            for name, nrs in self.phonebook_dict.items()
+                            for nr in nrs}
+        _LOGGER.info('Fritz!Box phone book successfully updated.')
+
+    def get_name(self, number):
+        """Return a name for a given phone number."""
+        number = re.sub(r'[^\d\+]', '', str(number))
+        if self.number_dict is None:
+            return 'unknown'
+        try:
+            return self.number_dict[number]
+        except KeyError:
+            pass
+        if self.prefixes:
+            for prefix in self.prefixes:
+                try:
+                    return self.number_dict[prefix + number]
+                except KeyError:
+                    pass
+                try:
+                    return self.number_dict[prefix + number.lstrip('0')]
+                except KeyError:
+                    pass
```

Comment:
Oh, and concerning closing the thread on HOMEASSISTANT_STOP, I'm sorry but I don't know how to do that. I tried to find other components where this is done but didn't find one. This concerns the part that is not modified in this PR.